src/integer_mod_q/mat_zq/inverse.rs

```diff
@@ -11,11 +11,65 @@
 use super::MatZq;
 use crate::{
     integer::{MatZ, Z},
-    traits::{Concatenate, GetNumRows},
+    traits::{Concatenate, Gcd, GetNumRows},
 };
 use flint_sys::fmpz_mod_mat::fmpz_mod_mat_rref;
 
 impl MatZq {
+    /// Returns the inverse of the matrix if it exists (is square and
+    /// has a determinant co-prime to the modulus) and `None` otherwise.
+    ///
+    /// # Examples
+    /// ```
+    /// use qfall_math::integer_mod_q::MatZq;
+    /// use std::str::FromStr;
+    ///
+    /// let mut matrix = MatZq::from_str("[[1,2],[3,4]] mod 7").unwrap();
+    ///
+    /// let matrix_invert = matrix.inverse().unwrap();
+    ///
+    /// let id = &matrix_invert * &matrix;
+    /// assert_eq!("[[5, 1],[5, 3]] mod 7", matrix_invert.to_string());
+    /// assert_eq!("[[1, 0],[0, 1]] mod 7", id.to_string());
+    /// ```
+    pub fn inverse(&self) -> Option<MatZq> {
+        // Check if the matrix is square and compute the determinant.
+        if let Ok(det) = MatZ::from(self).det() {
+            // Check weather the square matrix is invertible or not.
+            if det.gcd(self.get_mod()) != Z::ONE {
+                None
+            } else {
+                let dimensions = self.get_num_rows();
+                let mut inverse = MatZq::new(dimensions, dimensions, self.get_mod());
+
+                // The i`th unit vector in the for loop.
+                let mut e_i = MatZq::identity(dimensions, 1, self.get_mod());
+
+                // Use solve for all unit vectors except the last one.
+                for i in 0..dimensions - 1 {
+                    if let Some(column_i) = self.solve_gaussian_elimination(&e_i) {
+                        inverse.set_column(i, &column_i, 0).unwrap();
+                    } else {
+                        return None;
+                    }
+
+                    e_i.swap_entries(i, 0, i + 1, 0).unwrap();
+                }
+
+                // Use solve for the last unit vector.
+                if let Some(column_i) = self.solve_gaussian_elimination(&e_i) {
+                    inverse.set_column(dimensions - 1, &column_i, 0).unwrap();
+                } else {
+                    return None;
+                }
```

Comment:
why did you not include it in the loop ?

Reply:
Because swap_entries should not be computed after the last vector, so I could either test in every loop if we are on the last vector or I could exclude the computation for the last vector like above.
Since the latter is more efficient (but has more code) I chose that implementation

Reply:
considering the runtime of solve, and the very fast swap entries runtime, this seems like a very very low overhead, I would probably change it, even though it is slightly more ineffcient.
But do as you prefer in that case, the other just looks cleaner